File: src/db/store.ts
import { connect } from "@tursodatabase/database"
import {
  ALL_SCHEMA,
  ftsTableName,
  createFtsTableSQL,
  createFtsIndexSQL,
  dropFtsTableSQL,
  populateFtsSQL,
} from "./schema.ts"
import type { CodeChunk } from "../chunk/types.ts"

export interface SearchResult {
  chunkKey: string
  filePath: string
  name: string
  kind: string
  signature: string
  snippet: string
  startLine: number
  endLine: number
  score: number
}

export interface IndexedFile {
  filePath: string
  fileHash: string
  chunkCount: number
  indexedAt: number
}

export interface Codebase {
  id: number
  rootPath: string
  name: string
  indexedAt: number
  fileCount: number
  chunkCount: number
}

export class Store {
  private db: Awaited<ReturnType<typeof connect>>
  private initialized = false

  private constructor(db: Awaited<ReturnType<typeof connect>>) {
    this.db = db
  }

  static async open(dbPath: string): Promise<Store> {
    const db = await connect(dbPath, {
      experimental: ["index_method"],
    })
    const store = new Store(db)
    await store.init()
    return store
  }

  private async init(): Promise<void> {
    if (this.initialized) return
    for (const sql of ALL_SCHEMA) {
      await this.db.exec(sql)
    }
    this.initialized = true
  }

  // ── Codebase management ──────────────────────────────────────────

  /** Get or create a codebase entry, returns its id */
  async getOrCreateCodebase(rootPath: string, name?: string): Promise<number> {
    const row = await this.db.prepare(
      "SELECT id FROM codebases WHERE root_path = ?"
    ).get(rootPath) as { id: number } | undefined

    if (row) return row.id

    await this.db.prepare(
      "INSERT INTO codebases (root_path, name, indexed_at) VALUES (?, ?, ?)"
    ).run(rootPath, name ?? rootPath.split("/").pop() ?? "", Date.now())

    const inserted = await this.db.prepare(
      "SELECT id FROM codebases WHERE root_path = ?"
    ).get(rootPath) as { id: number }
    return inserted.id
  }

  /** List all codebases */
  async listCodebases(): Promise<Codebase[]> {
    const rows = await this.db.prepare(
      `SELECT c.id, c.root_path, c.name, c.indexed_at,
              COUNT(DISTINCT f.file_path) as file_count,
              COALESCE(SUM(f.chunk_count), 0) as chunk_count
       FROM codebases c
       LEFT JOIN indexed_files f ON f.codebase_id = c.id
       GROUP BY c.id
       ORDER BY c.root_path`
    ).all() as any[]
    return rows.map(r => ({
      id: r.id,
      rootPath: r.root_path,
      name: r.name,
      indexedAt: r.indexed_at,
      fileCount: r.file_count,
      chunkCount: r.chunk_count,
    }))
  }

  /** Update codebase indexed_at timestamp */
  async touchCodebase(codebaseId: number): Promise<void> {
    await this.db.prepare(
      "UPDATE codebases SET indexed_at = ? WHERE id = ?"
    ).run(Date.now(), codebaseId)
  }

  // ── File hash management ─────────────────────────────────────────

  /** Get stored file hash, or null if not indexed */
  async getFileHash(codebaseId: number, filePath: string): Promise<string | null> {
    const row = await this.db.prepare(
      "SELECT file_hash FROM indexed_files WHERE codebase_id = ? AND file_path = ?"
    ).get(codebaseId, filePath) as { file_hash: string } | undefined
    return row?.file_hash ?? null
  }

  // ── Chunk writes ─────────────────────────────────────────────────

  /** Batch insert chunks for multiple files in one transaction */
  async batchUpsertAllFileChunks(
    codebaseId: number,
    fileChunks: { filePath: string; fileHash: string; chunks: CodeChunk[] }[],
  ): Promise<void> {
    await this.db.exec("BEGIN")
    try {
      const now = Date.now()
      const deleteStmt = await this.db.prepare(
        "DELETE FROM chunks WHERE codebase_id = ? AND file_path = ?"
      )
      const insertStmt = await this.db.prepare(`
        INSERT INTO chunks (codebase_id, file_path, chunk_key, language, kind, name, signature, snippet, start_line, end_line, file_hash, indexed_at)
        VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
        ON CONFLICT(chunk_key) DO UPDATE SET
          codebase_id = excluded.codebase_id,
          language = excluded.language,
          kind = excluded.kind,
          name = excluded.name,
          signature = excluded.signature,
          snippet = excluded.snippet,
          start_line = excluded.start_line,
          end_line = excluded.end_line,
          file_hash = excluded.file_hash,
          indexed_at = excluded.indexed_at,
          embedding = NULL,
          embedding_model = ''
      `)
      const fileStmt = await this.db.prepare(`
        INSERT INTO indexed_files (codebase_id, file_path, file_hash, chunk_count, indexed_at)
        VALUES (?, ?, ?, ?, ?)
        ON CONFLICT(codebase_id, file_path) DO UPDATE SET
          file_hash = excluded.file_hash,
          chunk_count = excluded.chunk_count,
          indexed_at = excluded.indexed_at
      `)

      for (const { filePath, fileHash, chunks } of fileChunks) {
        await deleteStmt.run(codebaseId, filePath)
        for (const chunk of chunks) {
          await insertStmt.run(
            codebaseId,
            chunk.filePath,
            chunk.chunkKey,
            chunk.language,
            chunk.kind,
            chunk.name,
            chunk.signature,
            chunk.snippet,
            chunk.startLine,
            chunk.endLine,
            chunk.fileHash,
            now,
          )
        }
        await fileStmt.run(codebaseId, filePath, fileHash, chunks.length, now)
      }
      await this.db.exec("COMMIT")
    } catch (e) {
      await this.db.exec("ROLLBACK")
      throw e
    }
  }

  /** Remove all chunks and file records for a codebase */
  async removeCodebaseChunks(codebaseId: number): Promise<void> {
    await this.db.prepare("DELETE FROM chunks WHERE codebase_id = ?").run(codebaseId)
    await this.db.prepare("DELETE FROM indexed_files WHERE codebase_id = ?").run(codebaseId)
  }

  /** Remove files that are no longer on disk (scoped to codebase) */
  async removeStaleFiles(codebaseId: number, activeFiles: Set<string>): Promise<number> {
    const all = await this.db.prepare(
      "SELECT file_path FROM indexed_files WHERE codebase_id = ?"
    ).all(codebaseId) as { file_path: string }[]
    let removed = 0
    await this.db.exec("BEGIN")
    try {
      for (const row of all) {
        if (!activeFiles.has(row.file_path)) {
          await this.db.prepare(
            "DELETE FROM chunks WHERE codebase_id = ? AND file_path = ?"
          ).run(codebaseId, row.file_path)
          await this.db.prepare(
            "DELETE FROM indexed_files WHERE codebase_id = ? AND file_path = ?"
          ).run(codebaseId, row.file_path)
          removed++
        }
      }
      await this.db.exec("COMMIT")
    } catch (e) {
      await this.db.exec("ROLLBACK")
      throw e
    }
    return removed
  }

  // ── Embedding management ─────────────────────────────────────────

  /** Batch store embeddings */
  async batchUpsertEmbeddings(items: { chunkKey: string; embedding: number[]; modelName: string }[]): Promise<void> {
    if (items.length === 0) return
    await this.db.exec("BEGIN")
    try {
      for (const item of items) {
        const json = JSON.stringify(item.embedding)
        await this.db.prepare(
          `UPDATE chunks SET embedding = vector8(?), embedding_model = ? WHERE chunk_key = ?`
        ).run(json, item.modelName, item.chunkKey)
      }
      await this.db.exec("COMMIT")
    } catch (e) {
      await this.db.exec("ROLLBACK")
      throw e
    }
  }

  async countStaleEmbeddings(codebaseId: number, modelName: string): Promise<number> {
    const row = await this.db.prepare(
      `SELECT COUNT(*) as cnt FROM chunks WHERE codebase_id = ? AND (embedding IS NULL OR embedding_model != ?)`
    ).get(codebaseId, modelName) as { cnt: number }
    return Number(row.cnt)
  }

  /** Get chunks that need (re-)embedding (scoped to codebase) */
  async getStaleEmbeddings(codebaseId: number, modelName: string, limit?: number): Promise<{ chunkKey: string; name: string; signature: string; filePath: string; kind: string; snippet: string }[]> {
    const sql = limit 
      ? `SELECT chunk_key, name, signature, file_path, kind, snippet FROM chunks
         WHERE codebase_id = ? AND (embedding IS NULL OR embedding_model != ?)
         LIMIT ${limit}`
      : `SELECT chunk_key, name, signature, file_path, kind, snippet FROM chunks
         WHERE codebase_id = ? AND (embedding IS NULL OR embedding_model != ?)`
    
    const rows = await this.db.prepare(sql).all(codebaseId, modelName) as any[]
    return rows.map(r => ({
      chunkKey: r.chunk_key,
      name: r.name,
      signature: r.signature,
      filePath: r.file_path,
      kind: r.kind,
      snippet: r.snippet,
    }))
  }

  // ── Per-codebase FTS lifecycle ───────────────────────────────────

  /** Drop and rebuild FTS table for a codebase */
  async rebuildFtsTable(codebaseId: number): Promise<void> {
    // Drop old table (and its index)
    await this.db.exec(dropFtsTableSQL(codebaseId))
    // Create fresh table
    await this.db.exec(createFtsTableSQL(codebaseId))
    // Populate from chunks
    await this.db.prepare(populateFtsSQL(codebaseId)).run(codebaseId)
    // Build FTS index
    await this.db.exec(createFtsIndexSQL(codebaseId))
    // Optimize FTS index for faster queries
    await this.db.exec(`OPTIMIZE INDEX idx_${ftsTableName(codebaseId)}`)
  }

  // ── Search ───────────────────────────────────────────────────────

  /** Vector search across all codebases (global) */
  async vectorSearch(queryEmbedding: number[], limit: number, includeSnippet: boolean): Promise<SearchResult[]> {
    const json = JSON.stringify(queryEmbedding)
    const sql = includeSnippet
      ? `SELECT chunk_key, file_path, name, kind, signature, snippet, start_line, end_line,
                vector_distance_cos(embedding, vector8(?)) AS distance
         FROM chunks
         WHERE embedding IS NOT NULL
         ORDER BY distance ASC
         LIMIT ?`
      : `SELECT chunk_key, file_path, name, kind, signature, start_line, end_line,
                vector_distance_cos(embedding, vector8(?)) AS distance
         FROM chunks
         WHERE embedding IS NOT NULL
         ORDER BY distance ASC
         LIMIT ?`

    const rows = await this.db.prepare(sql).all(json, limit) as any[]

    return rows.map((row) => ({
      chunkKey: row.chunk_key,
      filePath: row.file_path,
      name: row.name,
      kind: row.kind,
      signature: row.signature,
      snippet: includeSnippet ? row.snippet : "",
      startLine: row.start_line,
      endLine: row.end_line,
      score: 1 - (row.distance ?? 1),
    }))
  }

  /** FTS search across all codebases (queries each FTS table, merges results) */
  async ftsSearch(query: string, limit: number, includeSnippet: boolean): Promise<SearchResult[]> {
    // Find all codebase IDs that have FTS tables
    const codebases = await this.db.prepare("SELECT id FROM codebases").all() as { id: number }[]

    const allResults: SearchResult[] = []

    for (const { id } of codebases) {
      const table = ftsTableName(id)
      // Check if FTS table exists
      const exists = await this.db.prepare(
        "SELECT name FROM sqlite_master WHERE type='table' AND name=?"
      ).get(table) as { name: string } | undefined
      if (!exists) continue

      try {
        const scores = await this.db.prepare(
          `SELECT chunk_id, fts_score(name, signature, ?1) AS score
           FROM ${table}
           WHERE fts_match(name, signature, ?1)
           ORDER BY score DESC
           LIMIT ?`
        ).all(query, limit) as { chunk_id: number; score: number }[]

        if (scores.length === 0) continue

        for (const { chunk_id, score } of scores) {
          const dataSql = includeSnippet
            ? `SELECT chunk_key, file_path, name, kind, signature, snippet, start_line, end_line FROM chunks WHERE id = ?`
            : `SELECT chunk_key, file_path, name, kind, signature, start_line, end_line FROM chunks WHERE id = ?`

          const row = await this.db.prepare(dataSql).get(chunk_id) as any
          if (!row) continue

          allResults.push({
            chunkKey: row.chunk_key,
            filePath: row.file_path,
            name: row.name,
            kind: row.kind,
            signature: row.signature,
            snippet: includeSnippet ? row.snippet : "",
            startLine: row.start_line,
            endLine: row.end_line,
            score,
          })
        }
      } catch (e: any) {
        // Only ignore "no such table" / "no such index" errors (FTS not built yet)
        const msg = String(e?.message ?? e)
        if (!msg.includes("no such table") && !msg.includes("no such index")) {
          throw e
        }
      }
    }

    // Sort by score descending, take top limit
    allResults.sort((a, b) => b.score - a.score)
    return allResults.slice(0, limit)
  }

  /** Count chunks that have embeddings (i.e., are searchable) */
  async countEmbeddedChunks(): Promise<number> {
    const row = await this.db.prepare(
      "SELECT COUNT(*) as cnt FROM chunks WHERE embedding IS NOT NULL"
    ).get() as { cnt: number }
    return row.cnt
  }

  /** List all indexed files (optionally scoped to codebase) */
  async listFiles(codebaseId?: number): Promise<IndexedFile[]> {
    const sql = codebaseId != null
      ? "SELECT file_path, file_hash, chunk_count, indexed_at FROM indexed_files WHERE codebase_id = ? ORDER BY file_path"
      : "SELECT file_path, file_hash, chunk_count, indexed_at FROM indexed_files ORDER BY file_path"

    const rows = codebaseId != null
      ? await this.db.prepare(sql).all(codebaseId) as any[]
      : await this.db.prepare(sql).all() as any[]

    return rows.map((row) => ({
      filePath: row.file_path,
      fileHash: row.file_hash,
      chunkCount: row.chunk_count,
      indexedAt: row.indexed_at,
    }))
  }

  close(): void {
    this.db.close()
  }
}

File: codemogger/store.go
package codemogger

import (
	"context"
	"database/sql"
	"encoding/binary"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"time"

	_ "github.com/tursodatabase/go-libsql"
)

type staleEmbedding struct {
	ChunkKey  string
	Name      string
	Signature string
	FilePath  string
	Kind      string
	Snippet   string
}

type embeddingUpdate struct {
	ChunkKey  string
	Embedding []float32
	ModelName string
}

type store struct {
	db *sql.DB
}

func openStore(dbPath string) (*store, error) {
	if err := os.MkdirAll(filepath.Dir(dbPath), 0o755); err != nil {
		return nil, err
	}
	db, err := sql.Open("libsql", fmt.Sprintf("file:%s", dbPath))
	if err != nil {
		return nil, err
	}
	s := &store{db: db}
	if err := s.init(); err != nil {
		_ = db.Close()
		return nil, err
	}
	return s, nil
}

func (s *store) init() error {
	// statements := []string{
	// 	`PRAGMA journal_mode = WAL`,
	// 	`PRAGMA foreign_keys = ON`,
	// 	`PRAGMA busy_timeout = 5000`,
	// }
	// for _, stmt := range statements {
	// 	if _, err := s.db.Exec(stmt); err != nil {
	// 		return err
	// 	}
	// }
	for _, stmt := range baseSchema {
		if _, err := s.db.Exec(stmt); err != nil {
			return err
		}
	}
	return nil
}

func (s *store) getOrCreateCodebase(ctx context.Context, rootPath string) (int64, error) {
	var id int64
	err := s.db.QueryRowContext(ctx, `SELECT id FROM codebases WHERE root_path = ?`, rootPath).Scan(&id)
	if err == nil {
		return id, nil
	}
	if !errors.Is(err, sql.ErrNoRows) {
		return 0, err
	}

	now := time.Now().UnixMilli()
	name := baseName(rootPath)
	if _, err := s.db.ExecContext(ctx, `INSERT INTO codebases (root_path, name, indexed_at) VALUES (?, ?, ?)`, rootPath, name, now); err != nil {
		return 0, err
	}
	err = s.db.QueryRowContext(ctx, `SELECT id FROM codebases WHERE root_path = ?`, rootPath).Scan(&id)
	return id, err
}

func (s *store) listCodebases(ctx context.Context) ([]Codebase, error) {
	rows, err := s.db.QueryContext(ctx, `
		SELECT c.id, c.root_path, c.name, c.indexed_at,
		       COUNT(DISTINCT f.file_path) AS file_count,
		       COALESCE(SUM(f.chunk_count), 0) AS chunk_count
		FROM codebases c
		LEFT JOIN indexed_files f ON f.codebase_id = c.id
		GROUP BY c.id
		ORDER BY c.root_path`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []Codebase
	for rows.Next() {
		var item Codebase
		if err := rows.Scan(&item.ID, &item.RootPath, &item.Name, &item.IndexedAt, &item.FileCount, &item.ChunkCount); err != nil {
			return nil, err
		}
		out = append(out, item)
	}
	return out, rows.Err()
}

func (s *store) touchCodebase(ctx context.Context, codebaseID int64) error {
	_, err := s.db.ExecContext(ctx, `UPDATE codebases SET indexed_at = ? WHERE id = ?`, time.Now().UnixMilli(), codebaseID)
	return err
}

func (s *store) getFileHash(ctx context.Context, codebaseID int64, filePath string) (string, error) {
	var hash string
	err := s.db.QueryRowContext(ctx, `SELECT file_hash FROM indexed_files WHERE codebase_id = ? AND file_path = ?`, codebaseID, filePath).Scan(&hash)
	if errors.Is(err, sql.ErrNoRows) {
		return "", nil
	}
	return hash, err
}

func (s *store) batchUpsertAllFileChunks(ctx context.Context, codebaseID int64, fileChunks []struct {
	FilePath string
	FileHash string
	Chunks   []CodeChunk
}) error {
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer func() {
		if err != nil {
			_ = tx.Rollback()
		}
	}()

	now := time.Now().UnixMilli()
	deleteStmt, err := tx.PrepareContext(ctx, `DELETE FROM chunks WHERE codebase_id = ? AND file_path = ?`)
	if err != nil {
		return err
	}
	defer deleteStmt.Close()

	insertStmt, err := tx.PrepareContext(ctx, `
		INSERT INTO chunks (
			codebase_id, file_path, chunk_key, language, kind, name, signature, snippet,
			start_line, end_line, file_hash, indexed_at
		) VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
		ON CONFLICT(chunk_key) DO UPDATE SET
			codebase_id = excluded.codebase_id,
			file_path = excluded.file_path,
			language = excluded.language,
			kind = excluded.kind,
			name = excluded.name,
			signature = excluded.signature,
			snippet = excluded.snippet,
			start_line = excluded.start_line,
			end_line = excluded.end_line,
			file_hash = excluded.file_hash,
			indexed_at = excluded.indexed_at,
			embedding = NULL,
			embedding_model = ''`)
	if err != nil {
		return err
	}
	defer insertStmt.Close()

	fileStmt, err := tx.PrepareContext(ctx, `
		INSERT INTO indexed_files (codebase_id, file_path, file_hash, chunk_count, indexed_at)
		VALUES (?, ?, ?, ?, ?)
		ON CONFLICT(codebase_id, file_path) DO UPDATE SET
			file_hash = excluded.file_hash,
			chunk_count = excluded.chunk_count,
			indexed_at = excluded.indexed_at`)
	if err != nil {
		return err
	}
	defer fileStmt.Close()

	for _, item := range fileChunks {
		if _, err = deleteStmt.ExecContext(ctx, codebaseID, item.FilePath); err != nil {
			return err
		}
		for _, chunk := range item.Chunks {
			if _, err = insertStmt.ExecContext(
				ctx,
				codebaseID,
				chunk.FilePath,
				chunk.ChunkKey,
				chunk.Language,
				chunk.Kind,
				chunk.Name,
				chunk.Signature,
				chunk.Snippet,
				chunk.StartLine,
				chunk.EndLine,
				chunk.FileHash,
				now,
			); err != nil {
				return err
			}
		}
		if _, err = fileStmt.ExecContext(ctx, codebaseID, item.FilePath, item.FileHash, len(item.Chunks), now); err != nil {
			return err
		}
	}

	err = tx.Commit()
	return err
}

func (s *store) removeStaleFiles(ctx context.Context, codebaseID int64, activeFiles map[string]struct{}) (int, error) {
	rows, err := s.db.QueryContext(ctx, `SELECT file_path FROM indexed_files WHERE codebase_id = ?`, codebaseID)
	if err != nil {
		return 0, err
	}
	defer rows.Close()

	var all []string
	for rows.Next() {
		var filePath string
		if err := rows.Scan(&filePath); err != nil {
			return 0, err
		}
		all = append(all, filePath)
	}
	if err := rows.Err(); err != nil {
		return 0, err
	}

	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return 0, err
	}
	defer func() {
		if err != nil {
			_ = tx.Rollback()
		}
	}()

	removed := 0
	for _, filePath := range all {
		if _, ok := activeFiles[filePath]; ok {
			continue
		}
		if _, err = tx.ExecContext(ctx, `DELETE FROM chunks WHERE codebase_id = ? AND file_path = ?`, codebaseID, filePath); err != nil {
			return 0, err
		}
		if _, err = tx.ExecContext(ctx, `DELETE FROM indexed_files WHERE codebase_id = ? AND file_path = ?`, codebaseID, filePath); err != nil {
			return 0, err
		}
		removed++
	}

	err = tx.Commit()
	return removed, err
}

func encodeEmbedding(vec []float32) []byte {
	buf := make([]byte, len(vec)*4)
	for i, v := range vec {
		binary.LittleEndian.PutUint32(buf[i*4:], mathFloat32bits(v))
	}
	return buf
}

func decodeEmbedding(blob []byte) []float32 {
	if len(blob)%4 != 0 {
		return nil
	}
	out := make([]float32, len(blob)/4)
	for i := range out {
		out[i] = mathFloat32frombits(binary.LittleEndian.Uint32(blob[i*4:]))
	}
	return out
}

func (s *store) batchUpsertEmbeddings(ctx context.Context, items []embeddingUpdate) error {
	if len(items) == 0 {
		return nil
	}
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer func() {
		if err != nil {
			_ = tx.Rollback()
		}
	}()

	stmt, err := tx.PrepareContext(ctx, `UPDATE chunks SET embedding = vector8(?), embedding_model = ? WHERE chunk_key = ?`)
	if err != nil {
		return err
	}
	defer stmt.Close()

	for _, item := range items {
		if _, err = stmt.ExecContext(ctx, encodeEmbedding(item.Embedding), item.ModelName, item.ChunkKey); err != nil {
			return err
		}
	}

	err = tx.Commit()
	return err
}

func (s *store) countStaleEmbeddings(ctx context.Context, codebaseID int64, modelName string) (int, error) {
	var count int
	err := s.db.QueryRowContext(ctx, `SELECT COUNT(*) FROM chunks WHERE codebase_id = ? AND (embedding IS NULL OR embedding_model != ?)`, codebaseID, modelName).Scan(&count)
	return count, err
}

func (s *store) getStaleEmbeddings(ctx context.Context, codebaseID int64, modelName string, limit int) ([]staleEmbedding, error) {
	query := `SELECT chunk_key, name, signature, file_path, kind, snippet
		FROM chunks
		WHERE codebase_id = ? AND (embedding IS NULL OR embedding_model != ?)`
	args := []any{codebaseID, modelName}
	if limit > 0 {
		query += ` LIMIT ?`
		args = append(args, limit)
	}

	rows, err := s.db.QueryContext(ctx, query, args...)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []staleEmbedding
	for rows.Next() {
		var item staleEmbedding
		if err := rows.Scan(&item.ChunkKey, &item.Name, &item.Signature, &item.FilePath, &item.Kind, &item.Snippet); err != nil {
			return nil, err
		}
		out = append(out, item)
	}
	return out, rows.Err()
}

func (s *store) rebuildFTS(ctx context.Context) error {
	tx, err := s.db.BeginTx(ctx, nil)
	if err != nil {
		return err
	}
	defer func() {
		if err != nil {
			_ = tx.Rollback()
		}
	}()

	for _, stmt := range []string{dropFTSSQL, createFTSSQL, populateFTSSQL} {
		if _, err = tx.ExecContext(ctx, stmt); err != nil {
			return err
		}
	}

	err = tx.Commit()
	return err
}

func (s *store) vectorSearch(ctx context.Context, queryEmbedding []float32, limit int, includeSnippet bool) ([]SearchResult, error) {
	query := `SELECT chunk_key, file_path, name, kind, signature, start_line, end_line,
                vector_distance_cos(embedding, vector8(?)) AS distance
         FROM chunks
         WHERE embedding IS NOT NULL
         ORDER BY distance ASC
         LIMIT ?`
	if includeSnippet {
		query = `SELECT chunk_key, file_path, name, kind, signature, snippet, start_line, end_line,
                vector_distance_cos(embedding, vector8(?)) AS distance
         FROM chunks
         WHERE embedding IS NOT NULL
         ORDER BY distance ASC
         LIMIT ?`
	}

	rows, err := s.db.QueryContext(ctx, query)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	results := make([]SearchResult, 0, limit)
	for rows.Next() {
		var item SearchResult
		var blob []byte
		if includeSnippet {
			if err := rows.Scan(&item.ChunkKey, &item.FilePath, &item.Name, &item.Kind, &item.Signature, &item.Snippet, &item.StartLine, &item.EndLine, &blob); err != nil {
				return nil, err
			}
		} else {
			if err := rows.Scan(&item.ChunkKey, &item.FilePath, &item.Name, &item.Kind, &item.Signature, &item.StartLine, &item.EndLine, &blob); err != nil {
				return nil, err
			}
		}
		item.Score = cosineSimilarity(queryEmbedding, decodeEmbedding(blob))
		results = append(results, item)
	}
	if err := rows.Err(); err != nil {
		return nil, err
	}

	sort.Slice(results, func(i, j int) bool {
		return results[i].Score > results[j].Score
	})
	if limit > 0 && len(results) > limit {
		results = results[:limit]
	}
	return results, nil
}

func (s *store) ftsSearch(ctx context.Context, query string, limit int, includeSnippet bool) ([]SearchResult, error) {
	selectCols := `c.chunk_key, c.file_path, c.name, c.kind, c.signature, c.start_line, c.end_line`
	if includeSnippet {
		selectCols = `c.chunk_key, c.file_path, c.name, c.kind, c.signature, c.snippet, c.start_line, c.end_line`
	}

	sqlQuery := `SELECT ` + selectCols + `,
		-bm25(chunks_fts, 5.0, 3.0) AS score
		FROM chunks_fts
		JOIN chunks c ON c.id = chunks_fts.rowid
		WHERE chunks_fts MATCH ?
		ORDER BY bm25(chunks_fts, 5.0, 3.0)
		LIMIT ?`

	rows, err := s.db.QueryContext(ctx, sqlQuery, query, limit)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var results []SearchResult
	for rows.Next() {
		var item SearchResult
		if includeSnippet {
			if err := rows.Scan(&item.ChunkKey, &item.FilePath, &item.Name, &item.Kind, &item.Signature, &item.Snippet, &item.StartLine, &item.EndLine, &item.Score); err != nil {
				return nil, err
			}
		} else {
			if err := rows.Scan(&item.ChunkKey, &item.FilePath, &item.Name, &item.Kind, &item.Signature, &item.StartLine, &item.EndLine, &item.Score); err != nil {
				return nil, err
			}
		}
		results = append(results, item)
	}
	return results, rows.Err()
}

func (s *store) listFiles(ctx context.Context) ([]IndexedFile, error) {
	rows, err := s.db.QueryContext(ctx, `SELECT file_path, file_hash, chunk_count, indexed_at FROM indexed_files ORDER BY file_path`)
	if err != nil {
		return nil, err
	}
	defer rows.Close()

	var out []IndexedFile
	for rows.Next() {
		var item IndexedFile
		if err := rows.Scan(&item.FilePath, &item.FileHash, &item.ChunkCount, &item.IndexedAt); err != nil {
			return nil, err
		}
		out = append(out, item)
	}
	return out, rows.Err()
}

func (s *store) close() error {
	return s.db.Close()
}
